### PR TITLE
Linting

### PR DESCRIPTION
**Pull request by florianwechsung**
Once the pyadjoint changes are merged

### CODE REVIEW

File: examples/L2tracking/L2tracking_main.py [hunk 1]
```diff
@@ -4,46 +4,45 @@
...
+# save state variable evolution in file u.pvd
 e.solve()
 out = fd.File("u.pvd")
-cb = lambda: out.write(e.solution)
-cb()
 
-#create PDEconstrained objective functional
-J_ = L2trackingObjective(e, Q, cb=cb)
+# create PDEconstrained objective functional
+J_ = L2trackingObjective(e, Q, cb=lambda: out.write(e.solution))
```
Comment (APaganini) [inline examples/L2tracking/L2tracking_main.py:22]: Before, we called `cb()` explicitely to save the initial domain. I suppose this is not necessary anymore. Please confirm

File: fireshape/control.py [hunk 2]
```diff
@@ -261,12 +283,14 @@ def __init__(self, mesh, bbox, orders, levels, fixed_dims=[],
...
-                               [2,..,2] means that they go to zero with C^1 regularity
+        boundary_regularities: how fast the splines go to zero on the boundary
+                               for each dimension
+                               [0,..,0] : they don't go to zero
+                               [1,..,1] : they go to zero with C^0 regularity
+                               [2,..,2] : they go to zero with C^1 regularity
         """
-        self.boundary_regularities = [o-1 for o in orders] if boundary_regularities is None else boundary_regularities 
+        self.boundary_regularities = [o - 1 for o in orders] \
+            if boundary_regularities is None else boundary_regularities
```
Comment (APaganini) [inline fireshape/control.py:293]: We should `assert` that `boundary_regularities` is compatible with `orders`. E.g.: linear splines (`order = 2`) cannot have C^2 regularity.

File: fireshape/control.py [hunk 3]
```diff
@@ -483,18 +511,19 @@ def construct_full_interpolation_matrix(self, IFW):
         FullIFW = PETSc.Mat().create(self.comm)
         FullIFW.setType(PETSc.Mat.Type.AIJ)
         d = self.dim
-        free_dims = list(set(range(self.dim))-set(self.fixed_dims))
+        free_dims = list(set(range(self.dim)) - set(self.fixed_dims))
         dfree = len(free_dims)
-        ((lsize, gsize), (lsize_spline, gsize_spline)) = IFW.getSizes()
-        FullIFW.setSizes(((d*lsize, d*gsize), (dfree*lsize_spline, dfree*gsize_spline)))
+        ((ls, gs), (ls_spline, gs_spline)) = IFW.getSizes()
```
Comment (APaganini) [inline fireshape/control.py]: in  line `492` we still have `lisize` and `gsize` instead of `ls` and `gs`

File: fireshape/control.py [hunk 4]
```diff
@@ -570,20 +600,20 @@ def __init__(self, controlspace: ControlSpace, inner_product: InnerProduct,
         else:
             self.fun = None
 
-
     def from_first_derivative(self, fe_deriv):
         if self.boundary_extension is not None:
```
Comment (APaganini) [inline fireshape/control.py:605]: please add explanation for this method

File: fireshape/control.py [hunk 5]
```diff
@@ -570,20 +600,20 @@ def __init__(self, controlspace: ControlSpace, inner_product: InnerProduct,
...
+            self.boundary_extension.apply_adjoint_action(
+                residual_smoothed, residual_smoothed)
             residual_smoothed -= p1
             self.controlspace.restrict(residual_smoothed, self)
         else:
             self.controlspace.restrict(fe_deriv, self)
 
-
     def to_coordinatefield(self, out):
         self.controlspace.interpolate(self, out)
```
Comment (APaganini) [inline fireshape/control.py:619]: please add explanation for this method

File: fireshape/objective.py [hunk 6]
```diff
@@ -174,6 +175,7 @@ def __init__(self, J: Objective, e: PdeConstraint):
         super().__init__(J.Q, J.cb)
         self.J = J
         self.e = e
+        fda.pause_annotation()
```
Comment (APaganini) [inline fireshape/objective.py:180]: please add comment

File: fireshape/objective.py [hunk 7]
```diff
@@ -182,24 +184,44 @@ def value(self, x, tol):
...
-            self.e.solve_adjoint(self.J.scale * self.J.value_form())
-        except fd.ConvergenceError:
-            if self.cb is not None:
-                self.cb()
-            raise
+        if self.Q.update_domain(x):
+            try:
+                # We use pyadjoint to calculate adjoint and shape derivatives,
+                # in order to do this we need to "record a tape of the forward
+                # solve", pyadjoint will then figure out all necesary adjoints.
```
Comment (APaganini) [inline fireshape/objective.py]: typo `necesary`

File: examples/L2tracking/L2tracking_main.py (hunk 1 above)
Comment (florianwechsung) [reply examples/L2tracking/L2tracking_main.py:22]: turns out this was never needed in the first place, ROL calls `update` at the initial guess and then `cb` is called.

File: fireshape/pde_constraint.py [hunk 8]
```diff
@@ -1,52 +1,12 @@
-import firedrake as fd
-
-
 class PdeConstraint(object):
     """
     Base class for PdeConstrain.
```
Comment (APaganini) [inline fireshape/pde_constraint.py]: typo `PdeConstrain`

File: setup.py [hunk 9]
```diff
@@ -1,8 +1,9 @@
 from setuptools import setup
 try:
-    import firedrake
```
Comment (APaganini) [inline setup.py:3]: add `import  firedrake_adjoint`

File: tests/test_levelset.py [hunk 10]
```diff
@@ -5,10 +5,16 @@
 import ROL
 
 
-@pytest.mark.parametrize("dim", [2, 3])
-@pytest.mark.parametrize("inner_t", [fs.H1InnerProduct, fs.ElasticityInnerProduct, fs.LaplaceInnerProduct])
-@pytest.mark.parametrize("controlspace_t", [fs.FeControlSpace, fs.FeMultiGridControlSpace, fs.BsplineControlSpace])
-@pytest.mark.parametrize("use_extension", ["wo_ext", "w_ext", "w_ext_fixed_fim"])
+@pytest.mark.parametrize("dim",
```
Comment (APaganini) [inline tests/test_levelset.py]: no `newline` after `"dim"`

**Review by APaganini — CHANGES_REQUESTED** [review]

File: fireshape/control.py (hunk 2 above)
Comment (florianwechsung) [reply fireshape/control.py:293]: I agree, but I don't think that is related to this Pull request, we are just doing formatting here.

File: fireshape/control.py (hunk 4 above)
Comment (florianwechsung) [reply fireshape/control.py:605]: as above.

File: fireshape/control.py (hunk 5 above)
Comment (florianwechsung) [reply fireshape/control.py:619]: as above.

**Review by APaganini — APPROVED** [review]

**Pull request merged**